src/napari_matplotlib/scatter.py

```diff
@@ -1,39 +1,225 @@
-        else:
+        data, x_axis_name, y_axis_name = self._get_data()
+
+        if len(data) == 0:
+            # don't plot if there isn't data
+            return
+
+        if self._histogram_for_large_data and (
+            data[0].size > self._threshold_to_switch_to_histogram
+        ):
             self.axes.hist2d(
                 data[0].ravel(),
                 data[1].ravel(),
                 bins=100,
                 norm=mcolor.LogNorm(),
             )
-        self.axes.set_xlabel(self.layers[0].name)
-        self.axes.set_ylabel(self.layers[1].name)
+        else:
+            self.axes.scatter(data[0], data[1], alpha=self._marker_alpha)
+
+        self.axes.set_xlabel(x_axis_name)
+        self.axes.set_ylabel(y_axis_name)
+
+    def _get_data(self) -> Tuple[List[np.ndarray], str, str]:
+        """Get the plot data.
+
+        This must be implemented on the subclass.
+
+        Returns
+        -------
+        data : np.ndarray
+            The list containing the scatter plot data.
+        x_axis_name : str
+            The label to display on the x axis
+        y_axis_name: str
+            The label to display on the y axis
+        """
+        raise NotImplementedError
+
+
+class ScatterWidget(ScatterBaseWidget):
+    """
+    Widget to display scatter plot of two similarly shaped image layers.
+
+    If there are more than 500 data points, a 2D histogram is displayed instead
+    of a scatter plot, to avoid too many scatter points.
+    """
+
+    n_layers_input = 2
+
+    def __init__(
+        self,
+        napari_viewer: napari.viewer.Viewer,
+    ):
+        super().__init__(
+            napari_viewer,
+        )
+
+    def _get_data(self) -> Tuple[List[np.ndarray], str, str]:
+        """Get the plot data.
+
+        Returns
+        -------
+        data : List[np.ndarray]
+            List contains the in view slice of X and Y axis images.
+        x_axis_name : str
+            The title to display on the x axis
+        y_axis_name: str
+            The title to display on the y axis
+        """
+        data = [layer.data[self.current_z] for layer in self.layers]
+        x_axis_name = self.layers[0].name
+        y_axis_name = self.layers[1].name
+
+        return data, x_axis_name, y_axis_name
+
+
+class FeaturesScatterWidget(ScatterBaseWidget):
+    n_layers_input = 1
+
+    def __init__(
+        self,
+        napari_viewer: napari.viewer.Viewer,
+        key_selection_gui: bool = True,
+    ):
+        self._key_selection_widget = None
+        super().__init__(
+            napari_viewer,
+        )
+
+        if key_selection_gui is True:
+            self._key_selection_widget = magicgui(
+                self._set_axis_keys,
+                x_axis_key={"choices": self._get_valid_axis_keys},
+                y_axis_key={"choices": self._get_valid_axis_keys},
+                call_button="plot",
+            )
+            self.layout().addWidget(self._key_selection_widget.native)
```

Comment:
This is awesome - I've been trying to get my head around how to use magicgui to do something like this to create part of a GUI for a while!

Reply:
Cool! Let me know if you have questions - happy to chat!